narwhal/consensus/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::bullshark::Bullshark;
+use crate::consensus::ConsensusProtocol;
+use crate::consensus::ConsensusState;
+use crate::consensus_utils::make_consensus_store;
+use crate::metrics::ConsensusMetrics;
+use config::{AuthorityIdentifier, Committee, Stake};
+use fastcrypto::hash::Hash;
+use fastcrypto::hash::HashFunction;
+use prometheus::Registry;
+use rand::distributions::Bernoulli;
+use rand::distributions::Distribution;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use rand::Rng;
+use rand::SeedableRng;
+use std::collections::{BTreeSet, HashMap, HashSet, VecDeque};
+use std::num::NonZeroUsize;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+use test_utils::mock_certificate_with_rand;
+use test_utils::CommitteeFixture;
+#[allow(unused_imports)]
+use tokio::sync::mpsc::channel;
+use types::HeaderAPI;
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% chance of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
```

Comment:
it is not clear here what a node failure is?

Reply:
Are you referring to the need for comments? I thought that it would be useful to make things clear to whoever intend to use the tests.

Reply:
Ah got it, let me refine this a bit